File: python-server/services/image_service.py
from utils.gemini_image_generation_helper import GeminiImageGenerator
from utils.gemini_image_generation_helper import GeminiImageGenerator
from utils.azure_storage_helper import AzureStorageHelper
from repositories.course_repo import CourseRepository
from repositories.subject_repo import SubjectRepository
from sqlalchemy.orm import Session
import logging

logger = logging.getLogger(__name__)

class ImageService:
    def __init__(self, db: Session):
        self.course_repo = CourseRepository(db)
        self.subject_repo = SubjectRepository(db)
        self.azure_storage = AzureStorageHelper()
        
    def generate_course_image(self, course_id):
        """Generate and store a cover image for a course"""
        try:
            # Get course details
            course = self.course_repo.get_course_by_id(course_id)
            if not course:
                raise ValueError(f"Course not found: {course_id}")
                
            # Initialize the image generator using environment variables
            generator = GeminiImageGenerator()
            
            # Generate the image
            logger.info(f"Generating image for course '{course.name}' (ID: {course_id})")
            image_bytes = generator.generate_course_image(course.name, course.description)
            if not image_bytes:
                raise ValueError("Failed to generate image")
            
            logger.info(f"Image generated successfully, size: {len(image_bytes)} bytes")
            
            # Save a local copy for debugging
            local_debug_path = f"/tmp/course_{course_id}_image_debug.png"
            try:
                with open(local_debug_path, 'wb') as f:
                    f.write(image_bytes)
                logger.info(f"Debug image saved to {local_debug_path}")
            except Exception as save_err:
                logger.warning(f"Could not save debug image: {str(save_err)}")
            
            # Upload the image to Azure Storage
            image_path = f"courses/{course_id}/cover"
            logger.info(f"Uploading image to Azure Storage path: {image_path}")
            image_url = self.azure_storage.upload_image(image_bytes, image_path)
            logger.info(f"Image uploaded successfully, URL: {image_url}")
            
            # Delete old image if it exists
            old_image_url = self.course_repo.get_course_image_url(course_id)
            if old_image_url:
                logger.info(f"Deleting old image: {old_image_url}")
                self.azure_storage.delete_image(old_image_url)
            
            # Update the course with the new image URL
            logger.info(f"Updating course with new image URL: {image_url}")
            updated_course = self.course_repo.update_course_image(course_id, image_url)
            
            return updated_course.to_dict() if updated_course else None
            
        except Exception as e:
            logger.error(f"Error generating course image: {str(e)}")
            raise
            
    def generate_subject_image(self, course_id, subject_id):
        """Generate and store a cover image for a subject"""
        try:
            # Get subject details
            subject = self.subject_repo.get_subject_by_id(subject_id)
            if not subject:
                raise ValueError(f"Subject not found: {subject_id}")
                
            # Get course name for context
            course = self.course_repo.get_course_by_id(course_id)
            if not course:
                raise ValueError(f"Course not found: {course_id}")
                
            # Initialize the image generator using environment variables
            generator = GeminiImageGenerator()
            
            # Generate the image
            image_bytes = generator.generate_subject_image(subject.name, course.name)
            if not image_bytes:
                raise ValueError("Failed to generate image")
                
            # Upload the image to Azure Storage
            image_path = f"courses/{course_id}/subjects/{subject_id}/cover"
            image_url = self.azure_storage.upload_image(image_bytes, image_path)
            
            # Delete old image if it exists
            old_image_url = self.subject_repo.get_subject_image_url(subject_id)
            if old_image_url:
                self.azure_storage.delete_image(old_image_url)
            
            # Update the subject with the new image URL
            updated_subject = self.subject_repo.update_subject_image(subject_id, image_url)
            
            return updated_subject.to_dict() if updated_subject else None
            
        except Exception as e:
            logger.error(f"Error generating subject image: {str(e)}")
            raise
            
    def generate_images_for_subjects(self, course_id):
        """Generate images for all subjects in a course"""
        try:
            # Get all subjects for the course
            subjects = self.subject_repo.get_subjects_by_course_id(course_id)
            
            results = []
            for subject in subjects:
                try:
                    result = self.generate_subject_image(course_id, subject.id)
                    results.append(result)
                except Exception as e:
                    logger.error(f"Error generating image for subject {subject.id}: {str(e)}")
                    results.append({"id": subject.id, "error": str(e)})
                    
            return results
            
        except Exception as e:
            logger.error(f"Error generating images for subjects: {str(e)}")
            raise

    def check_image_url(self, image_url):
        """Check if an image URL is valid and accessible"""
        try:
            import requests
            from PIL import Image
            from io import BytesIO
            
            # Try to fetch the image
            response = requests.get(image_url, timeout=5)
            response.raise_for_status()
            
            # Try to open it as an image
            Image.open(BytesIO(response.content))
            
            # If we got here, it's a valid image
            return {"valid": True, "message": "Image is valid"}
        except Exception as e:
            logger.error(f"Image URL check failed: {str(e)}")
            return {"valid": False, "error": str(e)}

File: go-server/services/image_service.go
package services

import (
	"context"
	"errors"
	"fmt"
	"go-server/models"
	"go-server/repositories"
	"go-server/utils"
	"os"
	"path/filepath"

	"github.com/sirupsen/logrus"
	"gorm.io/gorm"
)

type ImageService interface {
	GenerateCourseImage(courseID uint) (string, error)
	GenerateSubjectImage(subjectID uint) (string, error) 
	GenerateCustomImage(prompt string) (string, error)
	UploadImage(imageData []byte, imagePath string) (string, error)
	DeleteImage(imagePath string) error
	ListImages(pathPrefix string) ([]string, error)
}

type ImageServiceImpl struct {
	courseRepo      repositories.CourseRepository
	subjectRepo     repositories.SubjectRepository
	azureStorage    *utils.AzureStorageHelper
	imageGenerator  *utils.GeminiImageGenerator
}

func NewImageService(courseRepo repositories.CourseRepository, subjectRepo repositories.SubjectRepository) ImageService {
	return &ImageServiceImpl{
		courseRepo:     courseRepo,
		subjectRepo:    subjectRepo,
		azureStorage:   utils.GetAzureStorageHelper(),
		imageGenerator: utils.NewGeminiImageGenerator(),
	}
}

// GenerateCourseImage generates and stores a cover image for a course
func (s *ImageServiceImpl) GenerateCourseImage(courseID uint) (string, error) {
	// Get course details
	course := &models.Course{}
	if err := s.courseRepo.GetByID(courseID, course); err != nil {
		if err == gorm.ErrRecordNotFound {
			return "", errors.New("course not found")
		}
		return "", fmt.Errorf("failed to get course: %w", err)
	}

	// Check if image generator is available
	if !s.imageGenerator.IsAvailable() {
		return "", errors.New("image generation not available - no API key configured")
	}

	// Check if Azure storage is available
	if !s.azureStorage.IsInitialized() {
		return "", errors.New("Azure storage not initialized - check configuration")
	}

	// Generate the image
	ctx := context.Background()
	logrus.Infof("Generating image for course '%s' (ID: %d)", course.Name, courseID)
	
	imageBytes, err := s.imageGenerator.GenerateCourseImage(ctx, course.Name, course.Description)
	if err != nil {
		return "", fmt.Errorf("failed to generate image: %w", err)
	}

	if len(imageBytes) == 0 {
		return "", errors.New("failed to generate image - no data received")
	}

	logrus.Infof("Image generated successfully, size: %d bytes", len(imageBytes))

	// Save a local debug copy if in debug mode
	if os.Getenv("DEBUG") == "true" {
		debugPath := fmt.Sprintf("/tmp/course_%d_image_debug.png", courseID)
		if err := os.WriteFile(debugPath, imageBytes, 0644); err != nil {
			logrus.Warnf("Could not save debug image: %v", err)
		} else {
			logrus.Infof("Debug image saved to %s", debugPath)
		}
	}

	// Upload the image to Azure Storage
	imagePath := fmt.Sprintf("courses/%d/cover", courseID)
	logrus.Infof("Uploading image to Azure Storage path: %s", imagePath)
	
	imageURL, err := s.azureStorage.UploadImage(imageBytes, imagePath)
	if err != nil {
		return "", fmt.Errorf("failed to upload image to Azure Storage: %w", err)
	}

	logrus.Infof("Image uploaded successfully, URL: %s", imageURL)

	// Update the course with the image URL
	course.ImageURL = &imageURL
	if err := s.courseRepo.Update(course); err != nil {
		logrus.Errorf("Failed to update course with image URL: %v", err)
		// Don't fail the operation, just log the error
	}

	return imageURL, nil
}

// GenerateSubjectImage generates and stores a cover image for a subject
func (s *ImageServiceImpl) GenerateSubjectImage(subjectID uint) (string, error) {
	// Get subject details
	subject := &models.Subject{}
	if err := s.subjectRepo.GetByID(subjectID, subject); err != nil {
		if err == gorm.ErrRecordNotFound {
			return "", errors.New("subject not found")
		}
		return "", fmt.Errorf("failed to get subject: %w", err)
	}

	// Get course details for context
	course := &models.Course{}
	if err := s.courseRepo.GetByID(subject.CourseID, course); err != nil {
		return "", fmt.Errorf("failed to get course for subject: %w", err)
	}

	// Check if image generator is available
	if !s.imageGenerator.IsAvailable() {
		return "", errors.New("image generation not available - no API key configured")
	}

	// Check if Azure storage is available
	if !s.azureStorage.IsInitialized() {
		return "", errors.New("Azure storage not initialized - check configuration")
	}

	// Generate the image
	ctx := context.Background()
	logrus.Infof("Generating image for subject '%s' (ID: %d)", subject.Name, subjectID)
	
	imageBytes, err := s.imageGenerator.GenerateSubjectImage(ctx, subject.Name, course.Name)
	if err != nil {
		return "", fmt.Errorf("failed to generate subject image: %w", err)
	}

	if len(imageBytes) == 0 {
		return "", errors.New("failed to generate subject image - no data received")
	}

	logrus.Infof("Subject image generated successfully, size: %d bytes", len(imageBytes))

	// Upload the image to Azure Storage
	imagePath := fmt.Sprintf("subjects/%d/cover", subjectID)
	logrus.Infof("Uploading subject image to Azure Storage path: %s", imagePath)
	
	imageURL, err := s.azureStorage.UploadImage(imageBytes, imagePath)
	if err != nil {
		return "", fmt.Errorf("failed to upload subject image to Azure Storage: %w", err)
	}

	logrus.Infof("Subject image uploaded successfully, URL: %s", imageURL)

	return imageURL, nil
}

// GenerateCustomImage generates an image based on a custom prompt
func (s *ImageServiceImpl) GenerateCustomImage(prompt string) (string, error) {
	// Check if image generator is available
	if !s.imageGenerator.IsAvailable() {
		return "", errors.New("image generation not available - no API key configured")
	}

	// Check if Azure storage is available
	if !s.azureStorage.IsInitialized() {
		return "", errors.New("Azure storage not initialized - check configuration")
	}

	// Generate the image
	ctx := context.Background()
	logrus.Infof("Generating custom image with prompt: %s", prompt)
	
	imageBytes, err := s.imageGenerator.GenerateCustomImage(ctx, prompt)
	if err != nil {
		return "", fmt.Errorf("failed to generate custom image: %w", err)
	}

	if len(imageBytes) == 0 {
		return "", errors.New("failed to generate custom image - no data received")
	}

	logrus.Infof("Custom image generated successfully, size: %d bytes", len(imageBytes))

	// Upload the image to Azure Storage
	imagePath := fmt.Sprintf("custom/%d", utils.GenerateUniqueID())
	logrus.Infof("Uploading custom image to Azure Storage path: %s", imagePath)
	
	imageURL, err := s.azureStorage.UploadImage(imageBytes, imagePath)
	if err != nil {
		return "", fmt.Errorf("failed to upload custom image to Azure Storage: %w", err)
	}

	logrus.Infof("Custom image uploaded successfully, URL: %s", imageURL)

	return imageURL, nil
}

// UploadImage uploads an image directly to Azure Storage
func (s *ImageServiceImpl) UploadImage(imageData []byte, imagePath string) (string, error) {
	// Check if Azure storage is available
	if !s.azureStorage.IsInitialized() {
		return "", errors.New("Azure storage not initialized - check configuration")
	}

	if len(imageData) == 0 {
		return "", errors.New("no image data provided")
	}

	logrus.Infof("Uploading image to path: %s, size: %d bytes", imagePath, len(imageData))
	
	imageURL, err := s.azureStorage.UploadImage(imageData, imagePath)
	if err != nil {
		return "", fmt.Errorf("failed to upload image: %w", err)
	}

	logrus.Infof("Image uploaded successfully, URL: %s", imageURL)
	return imageURL, nil
}

// DeleteImage deletes an image from Azure Storage
func (s *ImageServiceImpl) DeleteImage(imagePath string) error {
	// Check if Azure storage is available
	if !s.azureStorage.IsInitialized() {
		return errors.New("Azure storage not initialized - check configuration")
	}

	// Extract just the path from the full URL if needed
	if filepath.IsAbs(imagePath) {
		// Extract the blob path from the full URL
		// This is a simple extraction, might need improvement for complex cases
		parts := filepath.Split(imagePath)
		imagePath = parts[1]
	}

	logrus.Infof("Deleting image: %s", imagePath)
	
	if err := s.azureStorage.DeleteImage(imagePath); err != nil {
		return fmt.Errorf("failed to delete image: %w", err)
	}

	logrus.Infof("Image deleted successfully: %s", imagePath)
	return nil
}

// ListImages lists all images in a specific path
func (s *ImageServiceImpl) ListImages(pathPrefix string) ([]string, error) {
	// Check if Azure storage is available
	if !s.azureStorage.IsInitialized() {
		return nil, errors.New("Azure storage not initialized - check configuration")
	}

	logrus.Infof("Listing images with prefix: %s", pathPrefix)
	
	imageURLs, err := s.azureStorage.ListImages(pathPrefix)
	if err != nil {
		return nil, fmt.Errorf("failed to list images: %w", err)
	}

	logrus.Infof("Found %d images with prefix: %s", len(imageURLs), pathPrefix)
	return imageURLs, nil
}